sse2neon.h

```diff
@@ -680,7 +689,7 @@ FORCE_INLINE void _sse2neon_kadd_f32(float *sum, float *c, float y)
     *sum = t;
 }
 
-#if defined(__ARM_FEATURE_CRYPTO)
+#if defined(__aarch64__) && defined(__ARM_FEATURE_CRYPTO)
```

Comment:
I am not confident to use `__ARM_ARCH == 8` or `__aarch64__` for such pre-condition.
Can you unify the convention?

Reply:
The code that follows that uses intrinsics that are only available on aarch64, so it needs that test.  32-bit ARMv8 targets can support CRC (most do, that's why clang enables it by default), and will define __ARM_FEATURE_CRYPTO with the correct compiler options (-mfpu=crypto-neon-fp-armv8), and thus need to be prevented from trying to compile that code.

Reply:
Ok, I see what's going on.  The support for the polynomial intrinsics that use poly64_t types is still missing for AArch32 in clang 12.  vreinterpret_p64_u64, vget_lane_p64, and vmull_p64 are all unavailable.

GCC supports them on AArch32, at least back through version 7.5.  So that code should be enabled for \_\_aarch64__, and AArch32 when compiled with GCC.  Since only ARMv8 and higher can set __ARM_FEATURE_CRYPTO, it isn't necessary to test for ARMv8 so that line could read:

`#if (!defined(__clang__) || defined(__aarch64__)) && defined(__ARM_FEATURE_CRYPTO)`

Reply:
One more thing: I'm not sure the __ARM_FEATURE_CRYPTO test is correct or even needed here.  That macro indicates support for the AES/SHA1/SHA256/SHA512 Neon intrinsics.  The code that is inside that test doesn't use any of those.  It really needs to be wrapped inside a test for poly64_t polynomial intrinsics support.  That support is supported in baseline ARMv8 for both 32-bit and 64-bit, but toolchain support only seems complete for AArch64, and AArch32 using GCC.

Reply:
It looks like __ARM_FEATURE_CRYPTO is required because of what appears to be an error in GCC's arm_neon.h for 32-bit ARM.  The poly64_t stuff is wrapped with  #pragma GCC target ("fpu=crypto-neon-fp-armv8"), so if you don't enable crypto support on the command line (which then defines __ARM_FEATURE_CRYPTO), you'll get a compilation error.  So the inclusion of __ARM_FEATURE_CRYPTO almost looks like something that was needed to work around this problem.

GCC's arm_neon.h for aarch64 does not do this - it doesn't care whether you've specified crypto support or not.

So to make this behavior match what it was before, but also enable it for GCC targeting AArch32, I might do something like:

```
#if defined(__ARM_FEATURE_CRYPTO) && \
   (defined(__aarch64__) || __has_builtin(__builtin_arm_crypto_vmullp64))
```

Reply:
Exactly, you can even add more comments around polynomial intrinsics.
Should the original `vmull_p8` fallback for A32 targets without polynomial intrinsics (clang?) be modified as well?

Reply:
vzip1q_u64 and vzip2q_u64 are only available on A64, so that code needs to remain conditional on __aarch64__.

I also confirmed that GCC's arm_neon.h for 64-bit ARM places a similar (seemingly unnecessary) requirement on vmull_p64 to compile with an option that enables crypto, so the test for __ARM_FEATURE_CRYPTO needs to stay the way it is.